src/ray/gcs/gcs_server/object_locator.h

```diff
@@ -0,0 +1,164 @@
+
+  std::unordered_set<ClientID> locations_;
+};
+
+/// \class NodeHoldObjectInfo
+/// This class is used to access the object ids which held by node.
+class NodeHoldObjectInfo {
+ public:
+  NodeHoldObjectInfo(const ClientID &node_id);
+
+  ~NodeHoldObjectInfo();
+
+  /// Add the id of the object that the node holds.
+  ///
+  /// \param object_id The id of the object that will be added.
+  void AddObject(const ObjectID &object_id);
+
+  /// Add the id of the object that the node holds.
+  ///
+  /// \param object_ids The ids of objects that will be added.
+  void AddObjects(const std::unordered_set<ObjectID> &object_ids);
+
+  /// Get all object of the node .
+  std::unordered_set<ObjectID> GetObjects();
+
+  /// Remove the id of the object that the node no longer holds.
+  ///
+  /// \param object_id The id of the object that will be removed.
+  /// \return The number of objects after remove.
+  size_t RemoveObject(const ObjectID &object_id);
+
+ private:
+  ClientID node_id_;
+
+  std::unordered_set<ObjectID> object_ids_;
+};
+
+class ObjectLocator {
+ public:
+  ObjectLocator();
+
+  ~ObjectLocator();
+
+  /// Add location of objects.
+  ///
+  /// \param node_id The object location that will be added.
+  /// \param object_ids The ids of objects which location will be added.
+  void AddObjectsLocation(const ClientID &node_id,
+                          const std::unordered_set<ObjectID> &object_ids);
+
+  /// Add location of an object.
+  ///
+  /// \param object_id The id of object which location will be added.
+  /// \param node_id The object location that will be added.
+  void AddObjectLocation(const ObjectID &object_id, const ClientID &node_id);
+
+  /// Get object's locations.
+  ///
+  /// \param object_id The id of object to lookup.
+  /// \return Object locations.
+  std::unordered_set<ClientID> GetObjectLocations(const ObjectID &object_id);
+
+  /// Remove object. This object will not be used again.
+  /// This will remove object from the object to location map and also the
+  /// remove from the node to objects map.
+  ///
+  /// \param object_id The id of object to be removed.
+  void RemoveObject(const ObjectID &object_id);
+
+  /// Remove the location from objects.
+  ///
+  /// \param node_id The location that will be removed.
+  void RemoveLocation(const ClientID &node_id);
+
+  /// Remove object's location.
+  ///
+  /// \param object_id The id of the object which location will be removed.
+  /// \param node_id The location that will be removed.
+  void RemoveObjectLocation(const ObjectID &object_id, const ClientID &node_id);
+
+ private:
+  /// Get ObjectLocationInfo by object id from map.
+  /// Will create it if not exist and the flag create_if_not_exist is set to true.
+  ///
+  /// \param object_id The id of object to lookup.
+  /// \param create_if_not_exist Whether to create a new one if not found.
+  /// \return std::shared_ptr<ObjectLocationInfo>
+  std::shared_ptr<ObjectLocationInfo> GetObjectLocationInfo(
```

Comment:
These private methods seem to just duplicate logic from the stdlib. Please remove them and just access the maps directly.

Reply:
How to implement this  "create_if_not_exist" function with one line of code? 
We use it more than once, so it's best to wrap it as a function to avoid duplicate code.

Reply:
Okay, that seems fine, but I would at least remove the deletion methods. You can just implement those with `erase`.

Reply:
Sure, I will remove the deletion methods.